### PR TITLE
[POC] Store file id in extended attributes where possible

### PR DESCRIPTION
**Pull request by icewind1991**
Requires http://pecl.php.net/package/xattr to be installed and the filesystem to support user extended attributes (default on btrfs, mount option for ext(3|4) iirc)

To test:
- upload a new file to a folder
- share the file
- delete the cache entry for the file
- check that the file is no longer shown in the web interface
- run the filescanner
- check that the file is back in the web interface and still shared

cc @PVince81 @DeepDiver1975

### CODE REVIEW

File: lib/private/files/cache/scanner.php [hunk 1]
```diff
@@ -213,12 +213,27 @@ protected function removeFromCache($path) {
 	protected function addToCache($path, $data, $fileId = -1) {
 		\OC_Hook::emit('Scanner', 'addToCache', array('file' => $path, 'data' => $data));
 		$this->emit('\OC\Files\Cache\Scanner', 'addToCache', array($path, $this->storageId, $data));
+		if ($this->storage->instanceOfStorage('OCP\Files\Storage\IAttribute') and $fileId === -1) {
```
Comment (PVince81) [inline lib/private/files/cache/scanner.php]: Does it mean you need to add `getAttribute` and `setAttribute` to the default storage wrapper, to make sure they are forwarded to the underlying storage ?

**Comment by PVince81** [comment]
Sweet.

We should probably also check for file id clashes. What happens if two files have the same file id attribute ? (for whatever reason)

File: lib/private/files/cache/scanner.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/files/cache/scanner.php]: No, there are magic methods to handle that

**Comment by MorrisJobke** [comment]
@icewind1991 rebase needed

**Comment by PVince81** [comment]
I wonder if we should do this as an optional separate app and register a storage wrapper or something ? This requires the FS to support xattr and the FS must be local, so no external storage (except the one of type "Local").

**Comment by icewind1991** [comment]
Some external storages support adding arbitrary metadata to files which could be used.

I'm not sure if this can be cleanly implemented in an app due to the changes to the scanner logic (which apps cant overwrite without hacks)

**Comment by PVince81** [comment]
Would it make sense to move this to an app instead ?

**Comment by DeepDiver1975** [comment]
This could be an alternative implementation for the upcoming storage based cache Händler.i vote die close for the time being.

**Pull request closed without merging**